lsコマンドを作る5(-a・-r・-lオプションすべてを実装したlsコマンド)のls.rb プログラムを作成

a-kuroki-gs opened the pull request.

a-kuroki-gs committed:
lsコマンドを作る5(-a・-r・-lオプションすべてを実装したlsコマンド)のls.rbプログラムを作成

a-kuroki-gs committed:
lsコマンドを作る5(-a・-r・-lオプションすべてを実装したlsコマンド)のls.rbプログラムを作成

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -26,10 +26,10 @@
 }.freeze
 
-def print_block_count(input)
+def print_block_count(input, a_param)
```
`a_param` だとパっと見何のことかわからないので、 `display_all` とか意味がわかるような名前にしたほうがよいかなと思いました。

a-kuroki-gs replied:
`display_all`→すべて表示する→`a`オプションのこと、ということですね！
訂正しました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -95,24 +95,66 @@ def print_contents(content, max_length)
+  print_filename_array(filename_array, max_length_array)
+end
+
+def check_a_option(filename)
+  filename.start_with?('.')
+end
```
ここはメソッド化しなくてもよいかなと思いました。メソッド化することで逆に何をしているのかわかりにくくなっているように思います。

メソッドとして名前をつけることは一種の抽象化です。メリットとして、

- 処理を意味の単位で一段抽象的な処理としてまとめて、再利用できる
- タイプ数が減らせる
- メソッドの中身の処理が変わってもメソッドの入出力が同じならメソッドを呼び出す側は変更が不要

などがありますが、逆にデメリットとして、
  
- 意味の単位でまとまっていないと、用途が不明瞭で再利用ができない
- 処理を細かく追いたいときにはメソッドの中身を見ないといけないので、コードを読むときに視線が無駄に上下してしまう
- アルゴリズムが隠蔽されてしまい、計算効率がわかりにくい

というのがあり、いつでもメソッドにすればOK、というわけではありません。

a-kuroki-gs replied:
ここはメソッドにするか迷った部分でもあったので、解説を聞いてなるほどと思いました。
このままだとわかりにくい＆無駄なプログラムになってしまっていますね。

訂正しました。

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -95,24 +95,66 @@ def print_contents(content, max_length)
+  element.reject { |e| check_a_option(e) }
+end
+
+def r_option(element)
+  element.reverse
+end
```
ここもメソッド化する意味があまりないかなと思います。

a-kuroki-gs replied:
上と同様に訂正しました。

yoshitsugu reviewed (commented):
コメントしました。

a-kuroki-gs committed:
ls5の訂正を行いました

不要なメソッドを削除し、該当部分を書き換える
わかりにくい変数名を変更

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -26,10 +26,10 @@
   total_blocks = 0
   Dir.foreach(input) do |i|
-    next if i.start_with?('.')
+    next if display_all.nil? && i.start_with?('.')
```
`display_all` は `true` か `false` のみが渡せる、という前提で判定できるようにしたほうがこのメソッドの中だけ見たときにわかりやすいかなと思いました。

a-kuroki-gs replied:
> `display_all` は `true` か `false` のみが渡せる、という前提で

`.nil?`とするのが冗長だと解釈し、`!`演算子を先頭につけて訂正しました。

`!`演算子を先頭につける、という使い方をしたことがなく、
今回この手法を知ることができました。ありがとうございます！

yoshitsugu commented on 04.ls/ls.rb:
```diff
@@ -95,24 +95,58 @@ def print_contents(content, max_length)
+  print_filename_array(filename_array, max_length_array)
+end
+
+def a_option(element)
+  element.reject { |e| e.start_with?('.') }
+end
```
ここも意味が不明瞭なので、 `reject_hidden_files` など、何をしているメソッドなのかわかるようにしましょう。
(ここではドットはじまりのファイルを hidden fileと表現しました)

a-kuroki-gs replied:
わかりにくかったので変更しました。
また`element`も`filename`に変更しております。

yoshitsugu reviewed (commented):
コメントしました。

a-kuroki-gs committed:
ls5のメソッド名や条件式の修正を行いました

一部の条件式の .nil? を削除し先頭に ! をつける
分かりにくいメソッド名をreject_hidden_filesに変更
ファイルを引数として受け取った際の-aオプションの処理を削除

yoshitsugu reviewed (approved):
OKです〜
